host/src/attribute.rs

```diff
@@ -654,19 +656,48 @@ impl<T: FromGatt> Characteristic<T> {
             return Ok(());
         }
 
-        let mut tx = P::allocate().ok_or(Error::OutOfMemory)?;
-        let mut w = WriteCursor::new(tx.as_mut());
-        let (mut header, mut data) = w.split(4)?;
-        data.write(crate::att::ATT_HANDLE_VALUE_NTF)?;
-        data.write(self.handle)?;
-        data.append(value)?;
+        let uns = AttUns::Notify {
+            handle: self.handle,
+            data: value,
+        };
+        let pdu = gatt::assemble(connection, crate::att::AttServer::Unsolicited(uns))?;
+        connection.send(pdu).await;
+        Ok(())
+    }
 
-        header.write(data.len() as u16)?;
-        header.write(4_u16)?;
-        let total = header.len() + data.len();
+    /// Write a value to a characteristic, and indicate a connection with the new value of the characteristic.
+    ///
+    /// If the provided connection has not subscribed for this characteristic, it will not be sent an indication.
+    ///
+    /// If the characteristic does not support indications, an error is returned.
+    ///
+    /// This function does not block for the confirmation to the indication message.
+    pub async fn indicate<P: PacketPool>(
+        &self,
+        connection: &GattConnection<'_, '_, P>,
+        value: &T,
+    ) -> Result<(), Error> {
+        let value = value.as_gatt();
+        let server = connection.server;
+        server.set(self.handle, value)?;
 
-        let pdu = crate::pdu::Pdu::new(tx, total);
+        let cccd_handle = self.cccd_handle.ok_or(Error::NotFound)?;
+        let connection = connection.raw();
+        if !server.should_indicate(connection, cccd_handle) {
+            // No reason to fail?
+            return Ok(());
+        }
+
+        let uns = AttUns::Indicate {
```

Comment:
If the only difference between `notify` and `indicate` is `AttUns`, I suggest to merge the shared part to something like `send`? The duplicate code looks unnecessary

Reply:
oh I didn't notice there was `should_notify` and `should_indicate`, if so the current impl is just fine!